Build frontend and backend in parallel in CI

Backend and frontend can be built in parallel in GitHub Actions. After they are built, frontend assets could be copied into the built JAR by:

1. Building only the `sirius-web-frontend` JAR
2. Copying it into the `sirius-web-sample-application` JAR directly by using https://stackoverflow.com/a/14388993/4874344

This should speed up the CI by about 2x, since the two jobs that run sequentially ATM each take around 2 minutes. If they run in parallel, they will take 2 mins each, and then there will be another short job to integrate the results.

When trying to do that naively:

```sh
zip -r staging/sirius-web-application.jar BOOT-INF/lib
```

(`BOOT-INF/lib` contains `sirius-web-frontend-0.5.0.jar` with frontend assets inside)

there is the following exception when starting the application:

```
19:13 $ java -jar staging/sirius-web-application.jar --spring.profiles.active=dev
Exception in thread "main" java.lang.IllegalStateException: Failed to get nested archive for entry BOOT-INF/lib/sirius-web-frontend-0.5.0.jar
        at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchive(JarFileArchive.java:120)
        at org.springframework.boot.loader.archive.JarFileArchive$NestedArchiveIterator.adapt(JarFileArchive.java:274)
        at org.springframework.boot.loader.archive.JarFileArchive$NestedArchiveIterator.adapt(JarFileArchive.java:265)
        at org.springframework.boot.loader.archive.JarFileArchive$AbstractIterator.next(JarFileArchive.java:226)
        at org.springframework.boot.loader.ExecutableArchiveLauncher.createClassLoader(ExecutableArchiveLauncher.java:87)
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:55)
        at org.springframework.boot.loader.JarLauncher.main(JarLauncher.java:88)
Caused by: java.io.IOException: Unable to open nested jar file 'BOOT-INF/lib/sirius-web-frontend-0.5.0.jar'
        at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:306)
        at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:292)
        at org.springframework.boot.loader.archive.JarFileArchive.getNestedArchive(JarFileArchive.java:116)
        ... 6 more
Caused by: java.lang.IllegalStateException: Unable to open nested entry 'BOOT-INF/lib/sirius-web-frontend-0.5.0.jar'. It has been compressed and nested jar files must be stored without compression. Please check the mechanism used to create your executable jar file
        at org.springframework.boot.loader.jar.JarFile.createJarFileFromFileEntry(JarFile.java:332)
        at org.springframework.boot.loader.jar.JarFile.createJarFileFromEntry(JarFile.java:314)
        at org.springframework.boot.loader.jar.JarFile.getNestedJarFile(JarFile.java:303)
        ... 8 more
```

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -7,7 +7,7 @@ jobs:
     runs-on: ubuntu-20.04
     defaults:
       run:
-        working-directory: frontend
+        working-directory: frontend/
     steps:
       - uses: actions/checkout@v2
       - uses: actions/setup-node@v2
@@ -30,20 +30,14 @@ jobs:
           path: frontend/build/
 
   build_backend:
+    # Build the backend without frontend assets.
+    # This way this can run in parallel to building the frontend.
+    # Integrating the 2 artifacts into the final JAR happens in the `build_app` job.
     name: Build backend
     runs-on: ubuntu-20.04
     defaults:
       run:
-        working-directory: backend
-    # Building the server requires frontend assets
-    # It could be optimized - adding the frontend assets to a prebuilt JAR
-    # could happen in a separate stage. Building backend and frontend could happen
-    # in parallel.
-    # The sirius-web-frontend package would have to be rebuilt and replaced
-    # in the final sirius-web-sample-application jar
-    # See https://docs.oracle.com/javase/tutorial/deployment/jar/update.html
-    # and https://stackoverflow.com/a/14388993/4874344
-    needs: build_frontend
+        working-directory: backend/
     steps:
       - uses: actions/checkout@v2
       - name: Set up JDK 11
@@ -52,10 +46,6 @@ jobs:
           java-version: "11"
           distribution: "adopt"
           cache: maven
-      - uses: actions/download-artifact@v2
-        with:
-          name: frontend-assets
-          path: backend/sirius-web-frontend/src/main/resources/static/
       - name: Generate Maven settings.xml
         run: >
           ./scripts/generate-settings.sh ${{ github.actor }} ${{
@@ -67,7 +57,53 @@ jobs:
         run: |
           mkdir staging
           find sirius-web-sample-application/target -maxdepth 1 -name "*.jar" -not -name "*-sources.jar" -exec cp {} staging/sirius-web-application.jar \;
-      - uses: actions/upload-artifact@v2
+      - name: Upload the backend JAR
+        uses: actions/upload-artifact@v2
+        with:
+          name: sirius-web-backend
+          path: backend/staging/
+
+  build_app:
+    name: Build the final application
+    runs-on: ubuntu-20.04
+    needs:
+      - build_frontend
+      - build_backend
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 11
+        uses: actions/setup-java@v2
+        with:
+          java-version: "11"
+          distribution: "adopt"
+          cache: maven
+
+      - name: Download frontend assets
+        uses: actions/download-artifact@v2
+        with:
+          name: frontend-assets
+          path: backend/sirius-web-frontend/src/main/resources/static/
+      - name: Build only the frontend Maven module
+        # https://stackoverflow.com/a/3899772/4874344
+        run: ./mvnw package -pl sirius-web-frontend
+        working-directory: backend/
+
+      - run: mkdir staging
+      - name: Download the backend JAR
+        uses: actions/download-artifact@v2
+        with:
+          name: sirius-web-backend
+          path: staging/
+
+      - name: Inject frontend assets into the JAR
+        # Injecting files into JAR using zip: https://stackoverflow.com/a/4799569/4874344
+        run: |
+          mkdir -p BOOT-INF/lib
+          find backend/sirius-web-frontend/target -maxdepth 1 -name "*.jar" -not -name "*-sources.jar" -exec mv {} BOOT-INF/lib/ \;
+          zip -r --compression-method=store staging/sirius-web-application.jar BOOT-INF/lib
+
+      - name: Upload the final application JAR
+        uses: actions/upload-artifact@v2
         with:
           name: sirius-web-application
-          path: backend/staging/
+          path: staging/

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -95,6 +95,14 @@ jobs:
           name: sirius-web-backend
           path: staging/
 
+      - name: Remove previous individual artifacts
+        # Helps keep the list of artifacts clean. They are not needed anyway
+        uses: geekyeggo/delete-artifact@v1
+        with:
+          name: |
+            sirius-web-backend
+            frontend-assets
+
       - name: Inject frontend assets into the JAR
         # Injecting files into JAR using zip: https://stackoverflow.com/a/4799569/4874344
         run: |

--- a/scripts/copy-frontend.sh
+++ b/scripts/copy-frontend.sh
@@ -6,7 +6,7 @@ script_dir="$(cd -- "$(dirname -- "${BASH_SOURCE[0]}")" &>/dev/null && pwd)"
 
 repo_dir=$(realpath "$script_dir/..")
 
-frontend_destination_directory="$repo_dir/backend/sirius-web-frontend/src/main/resources/static"
+frontend_destination_directory="$repo_dir/backend/sirius-web-frontend/src/main/resources/static/"
 
 mkdir -p "$frontend_destination_directory"
-cp -R "$repo_dir/frontend/build/" "$frontend_destination_directory"
+cp -R "$repo_dir/frontend/build/." "$frontend_destination_directory"

